0014-http-ilp/0004-http-ilp.md

```diff
@@ -0,0 +1,245 @@
+
+
+### 1.3. Server admin enters credentials into the HTTP-ILP server module config
+
+Next, Ankita logs back into her server and edits a config file of the HTTP-ILP server module to enter the receiver address and secret she obtained:
+
+``` ini
+[spsp_creds]
+receiver_address=ankita+filepay@nexus.justmoon.com
+receiver_secret=fxPERNaS4FGlC8H7eg6UfYVlglmFynFc8nh5la9PBGM
+```
+
+Next, she restarts her server to load the new configuration.
+
+### 1.4. Paid HTTP server fetches receiver information
+
+When the HTTP-ILP server module loads up, it uses [Webfinger](https://tools.ietf.org/html/rfc7033) to look up the `receiver_address` from the configuration.
+
+###### Request
+``` http
+GET /.well-known/webfinger?uri=ankita+filepay@nexus.justmoon.com
+Host: nexus@justmoon.com
+```
+
+The wallet will generate an ILP address using its ledger prefix, the fixed term `~recv`, the username `ankita` and the receiver name `filepay`.
+
+* ILP Address: `us.nexus.ankita.~recv.filepay`
+
+Note that the method for mapping an SPSP address to an ILP address is totally arbitrary and not part of this standard. The wallet MAY do this any way that it wants. In practice, the wallet may choose an address not including the username for privacy reasons.
+
+###### Response
+``` http
+HTTP/1.1 200 OK
+Content-Type: application/json
+```
+``` js
+{
+  // TODO ...
+}
+```
+
+The HTTP-ILP server module now knows the ILP address and caches it in memory for a certain period (depending on the cache header of the above request.) The reason we had Ankita enter the SPSP address and not the ILP address is because the ILP address may change over time as ILP providers change their routing structure. Note that ILP may support multihoming (this is still being debated as of this writing), so there may also be more than one ILP address.
+
+This completes the setup process. The server is now ready to receive paid API requests.
+
+## 2. Client interaction
+
+###### Scenario
+
+**Marat** is a *graphic designer* who would like to upload a mockup image to share with a client. His friend recommended a tool which doesn't require any signup or configuration and uses his existing ILP account to pay for the storage and bandwidth fees.
+
+![Sequence Diagram](sequence.png)
+
+### 2.1. Client generates a server-specific token from the hostname and its local secret:
+
+The uploader tool contains an HTTP-ILP client module. This client module has already generated and locally saved a `client_secret` using 256 bits of cryptographically secure randomness.
+
+Before making the paid HTTP request, the HTTP-ILP client module generates a **token** using the `hostname` of the server it is about to make a request to:
+
+* Token: `HMAC(SHA256, client_secret, hostname)`
+
+### 2.1. Client attempts to call API using token
+
+The paid HTTP request is fired off:
+
+``` http
+POST /upload HTTP/1.1
+Host: myservice.example
+X-Pay-Token: 7y0SfeN7lCuq0GFF5UsMYZofIjJ7LrvPvsePVWSv450
```

Comment:
Shouldn't it just be `Pay-Token`? See [RFC 6648](https://tools.ietf.org/html/rfc6648) on deprecating the X prefix

Reply:
Yeah, current unhash implementation uses `Pay` and `Pay-Token`

Reply:
oh, good to know :) I'll update that in my tutorial as well, then.